### PR TITLE
Fix Softmax in onnx frontend

### PR DESCRIPTION
**Pull request by hlu1**
The axis default value is not set when the attribute itself is absent in the model. This PR fixed that. Tested on SqueezeNet from https://github.com/onnx/models/tree/master/squeezenet.

### CODE REVIEW

**Comment by hlu1** [comment]
@tqchen @zhreshold, can you review please? Thanks

File: nnvm/python/nnvm/frontend/onnx.py [hunk 1]
```diff
@@ -598,7 +612,7 @@ def _get_convert_map(opset):
         # 'Mean'
         'Clip': AttrCvt('clip', transforms={'min': 'a_min', 'max': 'a_max'}),
         # softmax default axis is different in onnx
-        'Softmax': AttrCvt('softmax', {'axis': ('axis', 1)}),
+        'Softmax': Softmax.get_converter(opset),
```
[the remark this answers is not in the document]
Comment (hlu1) [reply nnvm/python/nnvm/frontend/onnx.py]: The output is all 1s without the fix, see:
https://gist.github.com/hlu1/23e70c367ed98feda878509043eb2ba9

**Comment by tqchen** [comment]
please add a regression testcase(possibly by constructing the onnx graph manually) to prevent this from happening again https://docs.tvm.ai/contribute/code_review.html#ensure-test-coverage

File: nnvm/tests/python/frontend/onnx/test_graph.py [hunk 2]
```diff
@@ -13,13 +14,19 @@ def compare_graph(onnx_file, nnvm_sym, ishape):
     ishapes = {input_name: ishape}
     graph_attr.set_shape_inputs(g1, ishapes)
     graph_attr.set_shape_inputs(g2, ishapes)
-    g1 = g1.apply("InferShape").apply("SimplifyInference")
-    g2 = g2.apply("InferShape").apply("SimplifyInference")
+    # g1 = g1.apply("InferShape").apply("SimplifyInference")
+    # g2 = g2.apply("InferShape").apply("SimplifyInference")
```
Comment (hlu1) [inline nnvm/tests/python/frontend/onnx/test_graph.py]: If I don't comment out these two lines, the tests crash at shape inference:
```
(tvm_env) hlu:tvm hlu$ python /Users/hlu/github/tvm/nnvm/tests/python/frontend/onnx/test_graph.py
Traceback (most recent call last):
  File "/Users/hlu/github/tvm/nnvm/tests/python/frontend/onnx/test_graph.py", line 31, in <module>
    test_super_resolution_example()
  File "/Users/hlu/github/tvm/nnvm/tests/python/frontend/onnx/test_graph.py", line 23, in test_super_resolution_example
    compare_graph(fname, symbol, ishape=(1, 1, 224, 224))
  File "/Users/hlu/github/tvm/nnvm/tests/python/frontend/onnx/test_graph.py", line 18, in compare_graph
    g2 = g2.apply("InferShape").apply("SimplifyInference")
  File "/Users/hlu/github/tvm/nnvm/python/nnvm/graph.py", line 234, in apply
    check_call(_LIB.NNGraphApplyPasses(self.handle, npass, cpass, ctypes.byref(ghandle)))
  File "/Users/hlu/github/tvm/nnvm/python/nnvm/_base.py", line 75, in check_call
    raise NNVMError(py_str(_LIB.NNGetLastError()))
nnvm._base.NNVMError: [14:03:31] /Users/hlu/github/tvm/nnvm/src/top/nn/convolution.cc:62: Check failed: dshape.ndim() == 4U (0 vs. 4) 62: Input data should be 4D

Stack trace returned 9 entries:
[bt] (0) 0   libnnvm_compiler.dylib              0x000000011afe9100 dmlc::StackTrace() + 288
[bt] (1) 1   libnnvm_compiler.dylib              0x000000011afe8e9f dmlc::LogMessageFatal::~LogMessageFatal() + 47
[bt] (2) 2   libnnvm_compiler.dylib              0x000000011b0d4739 nnvm::top::Conv2DInferShape(nnvm::NodeAttrs const&, std::__1::vector<nnvm::TShape, std::__1::allocator<nnvm::TShape> >*, std::__1::vector<nnvm::TShape, std::__1::allocator<nnvm::TShape> >*) + 3097
[bt] (3) 3   libnnvm_compiler.dylib              0x000000011b02cdb4 nnvm::Graph nnvm::pass::(anonymous namespace)::InferAttr<nnvm::TShape, nnvm::pass::(anonymous namespace)::$_0::operator()(nnvm::Graph) const::'lambda'(nnvm::TShape const&), std::nullptr_t>(nnvm::Graph&&, nnvm::TShape, char const*, char const*, char const*, char const*, char const*, nnvm::pass::(anonymous namespace)::$_0::operator()(nnvm::Graph) const::'lambda'(nnvm::TShape const&), std::nullptr_t)::'lambda'(unsigned int, bool)::operator()(unsigned int, bool) const + 3156
[bt] (4) 4   libnnvm_compiler.dylib              0x000000011b02adca std::__1::__function::__func<nnvm::pass::(anonymous namespace)::$_0, std::__1::allocator<nnvm::pass::(anonymous namespace)::$_0>, nnvm::Graph (nnvm::Graph)>::operator()(nnvm::Graph&&) + 3866
[bt] (5) 5   libnnvm_compiler.dylib              0x000000011affda4f nnvm::ApplyPasses(nnvm::Graph, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) + 1407
[bt] (6) 6   libnnvm_compiler.dylib              0x000000011afe4dd6 NNGraphApplyPasses + 566
[bt] (7) 7   _ctypes.cpython-36m-darwin.so       0x000000010c0ed49f ffi_call_unix64 + 79
[bt] (8) 8   ???                                 0x00007ffee43e37a0 0x0 + 140732727703456
```
I'm very new to tvm. Do you guys know why?

Comment (yzhliu) [reply nnvm/tests/python/frontend/onnx/test_graph.py]: the error msg shows it is something wrong with `g2`, which is constructed by nnvm symbol and should has nothing to do with your changes?

Comment (hlu1) [reply nnvm/tests/python/frontend/onnx/test_graph.py]: @yzhliu, I rebased onto master and tested again. The error went away. Must be a transient issue.

File: nnvm/tests/python/frontend/onnx/test_graph.py [hunk 3]
```diff
@@ -18,8 +19,16 @@ def compare_graph(onnx_file, nnvm_sym, ishape):
...
 
 def test_super_resolution_example():
-    fname, symbol = super_resolution, super_resolution_sym
+    fname, symbol = "super_resolution.onnx", super_resolution_sym
     compare_graph(fname, symbol, ishape=(1, 1, 224, 224))
 
+def test_squeeze_net():
+    # Only works for model downloaded from
+    # https://github.com/onnx/models/tree/master/squeezenet
+    fname = "squeezenet1_1.onnx"
```
Comment (hlu1) [inline nnvm/tests/python/frontend/onnx/test_graph.py]: I tested with the model in model zoo first and got this error:
```
Traceback (most recent call last):
  File "/Users/hlu/github/tvm/nnvm/tests/python/frontend/onnx/test_graph.py", line 32, in <module>
    test_squeeze_net()
  File "/Users/hlu/github/tvm/nnvm/tests/python/frontend/onnx/test_graph.py", line 28, in test_squeeze_net
    compare_graph(fname, symbol, ishape=(1, 3, 224, 224))
  File "/Users/hlu/github/tvm/nnvm/tests/python/frontend/onnx/test_graph.py", line 19, in compare_graph
    graph_util.check_graph_equal(g1, g2)
  File "/Users/hlu/github/tvm/nnvm/python/nnvm/compiler/graph_util.py", line 94, in check_graph_equal
    raise ValueError("Graph compare error: " + err)
ValueError: Graph compare error: Number of nodes mismatch (170 v.s 118)
```
Then I tested with the model I downloaded from the onnx modelzoo (https://github.com/onnx/models/tree/master/squeezenet), the test passed. 
@zhreshold, do you mind to update the model in model_zoo? The graph is also different from the model in the onnx model zoo

**Comment by hlu1** [comment]
I added unit test for softmax and graph test for squeezenet_1.1. @yzhliu, @zhreshold, can any of you review? Thanks!

File: nnvm/tests/python/frontend/onnx/test_graph.py (hunk 3 above)
Comment (zhreshold) [reply nnvm/tests/python/frontend/onnx/test_graph.py]: Can you have a quick debug why the old onnx model is no longer working? Just to make sure we don't break the legacy model support.

**Comment by tqchen** [comment]
please resolve the conflict with master by rebase, @yzhliu @zhreshold can you followup?

**Comment by hlu1** [comment]
@zhreshold, sorry about the late reply. There is add add layer after every conv layer in the legacy squeezenet onnx model, which is transformed to expand_dims + broadcast_add after being converted to nnvm graph (see debug_str printout here: https://gist.github.com/hlu1/282e304a064a4315b69223492d672f27). There is also no softmax layer in the end, so making the legacy onnx model work won't be relevant to this PR. The current squeezenet model in the onnx model zoo has the same network layers as in https://github.com/DeepScale/SqueezeNet/blob/master/SqueezeNet_v1.1/deploy.prototxt, which has a softmax layer as the last layer and no add layer in the middle.

**Comment by zhreshold** [comment]
@hlu1 Thanks for the explanation , LGTM.

**Comment by tqchen** [comment]
Thanks @hlu1 this is now merged

**Pull request merged**